File: src/main/java/com/github/simiacryptus/aicoder/text/LineComment.java
package com.github.simiacryptus.aicoder.text;

import org.jetbrains.annotations.NotNull;

import java.util.Arrays;
import java.util.stream.Collectors;

public class LineComment extends IndentedText {
    public static class Factory implements TextBlockFactory<LineComment> {
        public final String commentPrefix;

        public Factory(String commentPrefix) {
            this.commentPrefix = commentPrefix;
        }

        @Override
        public LineComment fromString(String text) {
            text = text.replace("\t", TAB_REPLACEMENT);
            String indent = StringTools.getWhitespacePrefix(text.split(DELIMITER));
            return new LineComment(commentPrefix, indent, Arrays.stream(text.split(DELIMITER))
                    .map(s -> StringTools.stripPrefix(s, indent))
                    .map(StringTools::trimPrefix)
                    .map(s -> StringTools.stripPrefix(s, commentPrefix))
                    .toArray(String[]::new));
        }

        @Override
        public boolean looksLike(String text) {
            return Arrays.stream(text.split(DELIMITER)).allMatch(x->x.trim().startsWith(commentPrefix));
        }
    }

    private final String commentPrefix;

    public LineComment(String commentPrefix, String indent, String... textBlock) {
        super(indent, textBlock);
        this.commentPrefix = commentPrefix;
    }

    @Override
    public @NotNull String toString() {
        return commentPrefix + " " + Arrays.stream(rawString()).collect(Collectors.joining(DELIMITER + getIndent() + commentPrefix + " "));
    }

    public @NotNull IndentedText withIndent(String indent) {
        return new LineComment(commentPrefix, indent, textBlock);
    }
}

File: src/main/kotlin/com/simiacryptus/aicoder/util/LineComment.kt
package com.simiacryptus.aicoder.util

import com.simiacryptus.util.StringUtil.getWhitespacePrefix
import com.simiacryptus.util.StringUtil.stripPrefix
import com.simiacryptus.util.StringUtil.trimPrefix
import java.util.*
import java.util.stream.Collectors

class LineComment(private val commentPrefix: CharSequence, indent: CharSequence?, vararg lines: CharSequence) :
    IndentedText(indent!!, *lines) {
    class Factory(private val commentPrefix: String) : TextBlockFactory<LineComment?> {
        override fun fromString(text: String?): LineComment {
            var textVar = text
            textVar = textVar!!.replace(Regex("\t"), TextBlock.TAB_REPLACEMENT.toString())
            val indent =
                getWhitespacePrefix(*textVar.split(TextBlock.DELIMITER.toRegex()).dropLastWhile { it.isEmpty() }
                    .toTypedArray())
            return LineComment(
                commentPrefix,
                indent,
                *Arrays.stream(textVar.split(TextBlock.DELIMITER.toRegex()).dropLastWhile { it.isEmpty() }
                    .toTypedArray())
                    .map { s: String? ->
                        stripPrefix(
                            s!!, indent
                        )
                    }
                    .map { obj: CharSequence -> trimPrefix(obj) }
                    .map { s: CharSequence? ->
                        stripPrefix(
                            s!!,
                            commentPrefix
                        )
                    }
                    .collect(Collectors.toList()).toTypedArray())
        }

        override fun looksLike(text: String?): Boolean {
            return Arrays.stream(text!!.split(TextBlock.DELIMITER.toRegex()).dropLastWhile { it.isEmpty() }
                .toTypedArray()).allMatch { x: String ->
                x.trim { it <= ' ' }.startsWith(
                    commentPrefix
                )
            }
        }
    }

    override fun toString(): String {
        return "$commentPrefix " + Arrays.stream(rawString())
            .collect(Collectors.joining(TextBlock.DELIMITER + indent + commentPrefix + " "))
    }

    override fun withIndent(indent: CharSequence): LineComment {
        return LineComment(commentPrefix, indent, *lines)
    }
}
